### PR TITLE
refs #14 Rework of --output path and file name handling

### PR DESCRIPTION
**Pull request by lta14**
Reworked the file name handling and adapted the documentation accordingly.

### CODE REVIEW

File: antiweb.py [hunk 1]
```diff
@@ -2780,44 +2794,94 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
-                    out_file = os.path.splitext(fname)[0] + docs + ".rst"
-                else:
-                    out_file = os.path.splitext(fname)[0] + ".rst"
-                out_file_name = os.path.split(out_file)[1]
+                out_file_name_index = os.path.relpath(out_file_name_index, path)
 
 #@edoc
-#There is an output given, so it can also be either recursive or not recursive. With the additional output parameter the file path gets split up and put together so it can be processed by ''process_file''. There is also a differentiation between Linux, Windows and OS X (because of the different paths in each operating system)
+
+#If there is an output given, we have to distinguishe between the non-recursive and recursive option:
```
Comment (7enderhead) [inline antiweb.py]: 'distinguish'

Comment (7enderhead) [inline antiweb.py]: !fix

File: antiweb.py [hunk 2]
```diff
@@ -2780,44 +2794,94 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+
+#.. csv-table::
+#   :header: "Input File", "Output File"
+
+#   ``C:\antiweb\testing.cs -o report.rst``,*working_dir\\report.rst*
+#   ``C:\antiweb\testing.cs -o report``,*working_dir\\report\\testing.rst*
+#   ``C:\antiweb\testing.cs -o report\report.rst``,*C:\\antiweb\\report\\report.rst*
```
Comment (7enderhead) [inline antiweb.py]: What is the logic that when the options from the first two lines (which write to current working_dir) are combined that the output is now relative to the original source dir?

Comment (7enderhead) [inline antiweb.py]: !resolve

File: antiweb.py [hunk 3]
```diff
@@ -2780,44 +2794,94 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+#When the recursive option is used in combination with the output option, the output parameter is treated as the documentation directory:
+
+#.. csv-table::
+#   :header: "Input File", "Output File"
+
+#   ``C:\antiweb\ -r -o report.rst``,*working_dir\\report.rst\\file.rst*
+#   ``C:\antiweb\ -r -o report``,*working_dir\\report\\file.rst*
+#   ``C:\antiweb\ -r -o C:\antiweb\report.rst``,*C:\\antiweb\\report.rst\\file.rst*
+
+#The output file name for the index file is the file name itself: *testing.rst*.
```
Comment (7enderhead) [inline antiweb.py]: I don't understand this paragraph:
- Which 'filename' is used as the index file?
- The path separator part is unclear !resolve

File: antiweb.py [hunk 4]
```diff
@@ -2780,44 +2794,94 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
 
 #@edoc
```
Comment (7enderhead) [inline antiweb.py]: The parameters for function 'write' need documentation. !fix

File: antiweb.py [hunk 5]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+    """
+.. py:method:: get_output_file_name(working_dir, input_file_name, output, directory, file)
+    
+   Computes the absolute path to the output file based on the output commandline option.
+           
+   :param working_dir: Current working directory.
+   :param input_file_name: The name of the file which should be processed.
+   :param output: The given -output parameter.
+   :param directory: The dedicated report directory.
+   :param file: The file part from the -output parameter.
```
Comment (7enderhead) [inline antiweb.py]: Why do you need it as a separate parameter? Can you compute it internally from the _output_ parameter? !resolve

File: antiweb.py [hunk 6]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+    out_file = ""
+
+    if not directory:
+        if file:
+            #check if the file has an extension. 
+            #true: the output parameter is treated as the outputfile name
+            #otherwise it is treated as the documentation directory
+            file_name, file_extension = os.path.splitext(file)
+            if file_extension:
+                out_file = os.path.join(working_dir, directory, file_name + ".rst")
```
Comment (7enderhead) [inline antiweb.py]: Why do you join with _directory_, which does not seem to be set in this if branch? !resolve

File: antiweb.py (hunk 5 above)
Comment (7enderhead) [inline antiweb.py]: Also, the parameter name is not good (it is stronger than _input_file_name_). !fix

File: antiweb.py [hunk 7]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+            else:
+                #the output parameter refers to a directory. the directory is combined with
+                #the input file name and optionally with the working_dir.
+                out_file = create_output_file_name(working_dir, input_file_name,  file_name)
+    else:
+        #we got a directory from the -output parameter. if the output parameter only contains a directory
```
Comment (7enderhead) [inline antiweb.py]: Please use proper casing when writing full sentences. Use bullet points otherwise. !fix

File: antiweb.py [hunk 8]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
-                    replace_in_generated(start_of_block, end_of_block, out_file_name, path, output, index_rst, content, startline)
+                    replace_in_generated(start_of_block, end_of_block, out_file_name_index, path, output, index_rst, content, startline)
                 else:
-                    replace_in_generated(start_of_block, end_of_block, out_file_name, path, None, index_rst, content, startline)
+                    replace_in_generated(start_of_block, end_of_block, out_file_name_index, path, None, index_rst, content, startline)
 #@edoc
-
 #@(write)
+
+#@cstart(get_output_file_name)
```
Comment (7enderhead) [inline antiweb.py]: This function is sufficently confusing that it needs some examples. !fix

File: antiweb.py [hunk 9]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+
+#@cstart(create_output_file_name)
+
+def create_output_file_name(working_dir, input_file_name, directory):
+#@start(create_output_file_name doc)
+    """
+.. py:method:: create_output_file_name(working_dir, input_file_name, directory)
+   
+  Computes the path to the output file name. The ".rst" suffix is added to the input file name.
+  Then the paths of the working_dir and the directory are combined with the input file name.
```
Comment (7enderhead) [inline antiweb.py]: 'given directory' !fix

File: antiweb.py [hunk 10]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+             
+  :param working_dir: Current working directory.
+  :param input_file_name: The name of the file which should be processed.
+  :param directory: The dedicated report directory.
+  :return: The path of the output file.
```
Comment (7enderhead) [inline antiweb.py]: _Absoulte_ or _relative_? !fix

File: antiweb.py [hunk 11]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+#@start(create_output_file_name doc)
+    """
+.. py:method:: create_output_file_name(working_dir, input_file_name, directory)
+   
+  Computes the path to the output file name. The ".rst" suffix is added to the input file name.
+  Then the paths of the working_dir and the directory are combined with the input file name.
+             
+  :param working_dir: Current working directory.
+  :param input_file_name: The name of the file which should be processed.
+  :param directory: The dedicated report directory.
```
Comment (7enderhead) [inline antiweb.py]: So this is always treated as _relative_, right? If so, please document. !fix

File: antiweb.py [hunk 12]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+#@(create_output_file_name)
+
+#@cstart(create_doc_directory)
+
+def create_doc_directory(out_file):
+#@start(create_doc_directory doc)
+    """
+.. py:method:: create_doc_directory(out_file)
+    
+   Creates the documentation directory if it not yet exists.
```
Comment (7enderhead) [inline antiweb.py]: 'does not yet exit' !fix

Comment (7enderhead) [inline antiweb.py]: 'exist', of course

File: antiweb.py [hunk 13]
```diff
@@ -2825,21 +2889,112 @@ def write(path, fname, output, token, warnings, index, index_rst, recursive, con
...
+   If an error occurs the program exits.
+           
+   :param out_file: The path to the output file.
+    """
+#@include(create_doc_directory) 
+#@(create_doc_directory doc)        
+    try:
```
Comment (7enderhead) [inline antiweb.py:3046]: In the documentation output, the code for this function has line numbers, but the code sections for _get_output_file_name_ and _create_output_file_name_ do not. Why? !fix

File: antiweb.py [hunk 14]
```diff
@@ -2941,7 +3096,7 @@ def main():
                 content, startline = search_for_generate(os.getcwd(), index_rst, start_of_block, end_of_block)
 #@edoc
 
-#The program lists all files in the directory and sub-directories to prepare them for the process
+#The program walks through the given directory and all subdirectories and retrieves all files.
```
Comment (7enderhead) [inline antiweb.py]: It does not retrieve files, but _file names_ (absolute or relative?)... !fix

File: antiweb.py (hunk 4 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 13 above)
Comment (lta14) [inline antiweb.py:3046]: !fixed

File: antiweb.py (hunk 2 above)
Comment (lta14) [inline antiweb.py]: !resolved

File: antiweb.py (hunk 3 above)
Comment (lta14) [inline antiweb.py]: !resolved

File: antiweb.py (hunk 8 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 3 above)
Comment (lta14) [inline antiweb.py]: !resolved

File: antiweb.py (hunk 5 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 6 above)
Comment (lta14) [inline antiweb.py]: !resolved

File: antiweb.py (hunk 7 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 12 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 10 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 11 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 9 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 14 above)
Comment (lta14) [inline antiweb.py]: !fixed

File: antiweb.py (hunk 1 above)
Comment (lta14) [inline antiweb.py]: !fixed

**Pull request merged**